kv/kvnemesis: TestKVNemesisSingleNode failed

kv/kvnemesis.TestKVNemesisSingleNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9246422?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9246422?buildTab=artifacts#/) on release-23.1 @ [f351747ed97862fc037717cadec23f18073fb6be](https://github.com/cockroachdb/cockroach/commits/f351747ed97862fc037717cadec23f18073fb6be):


```
=== RUN   TestKVNemesisSingleNode
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisSingleNode1805654162
    test_log_scope.go:79: use -show-logs to present logs inline
    kvnemesis_test.go:180: seed: 6246002925880804455
    kvnemesis_test.go:124: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis4063208312
    kvnemesis.go:165: error applying x.AddSSTable(ctx, tk(1705712307504063096), tk(10247965761760221684), ... /* @s20 */) // 1377 bytes
        // ^-- [tk(1977974164538652706), tk(2045593780759141697)) -> sv(s20): /Table/100/"1{b732d05dc39f822"-c6368b4a0e59d41"} -> /<empty>
        // ^-- tk(2045593780759141697) -> sv(s20): /Table/100/"1c6368b4a0e59d41"/0,1 -> /BYTES/v20
        // ^-- tk(2356936398634376679) -> sv(s20): /Table/100/"20b5852abeca5de7"/0,1 -> /<empty>
        // ^-- tk(3295195921125409205) -> sv(s20): /Table/100/"2dbae33b47ff01b5"/0,1 -> /<empty>
        // ^-- tk(5582719925329745694) -> sv(s20): /Table/100/"4d79ce318dc6571e"/0,1 -> /BYTES/v20
        // ^-- tk(6899655393190266888) -> sv(s20): /Table/100/"5fc080062fb6c808"/0,1 -> /BYTES/v20
        // ^-- tk(10247965761760221683) -> sv(s20): /Table/100/"8e381690741559f3"/0,1 -> /BYTES/v20: checking for key collisions: expected engine iter to be ahead of sst iter
    kvnemesis.go:185: failures(verbose): /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis4063208312/failures
        repro steps: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis4063208312/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis4063208312/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis4063208312/kvs-scan.txt
    kvnemesis_test.go:207: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:207
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:148
        	Error:      	Should be zero, but was 1
        	Test:       	TestKVNemesisSingleNode
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisSingleNode1805654162
--- FAIL: TestKVNemesisSingleNode (10.98s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #99010 kv/kvnemesis: TestKVNemesisSingleNode failed [C-test-failure O-robot T-kv branch-master]
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisSingleNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-25939

--- a/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go
+++ b/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go
@@ -1083,6 +1083,12 @@ func TestEvalAddSSTable(t *testing.T) {
 			sst:          kvs{pointKV("oe", 11, "foo"), pointKV("oih", 12, "foo"), rangeKV("ods", "ogvh", 10, ""), rangeKV("ogvh", "ohl", 10, ""), rangeKV("ogvh", "ohl", 9, "")},
 			ignoreExpect: true,
 		},
+		"DisallowConflict maintains ext iter ahead of sst iter": {
+			noConflict:   true,
+			data:         kvs{pointKV("c", 6, "foo"), rangeKV("c", "e", 5, "")},
+			sst:          kvs{rangeKV("a", "b", 10, ""), pointKV("d", 9, "foo")},
+			ignoreExpect: true,
+		},
 	}
 	testutils.RunTrueAndFalse(t, "IngestAsWrites", func(t *testing.T, ingestAsWrites bool) {
 		testutils.RunValues(t, "RewriteConcurrency", []interface{}{0, 8}, func(t *testing.T, c interface{}) {

--- a/pkg/storage/sst.go
+++ b/pkg/storage/sst.go
@@ -1137,6 +1137,10 @@ func CheckSSTConflicts(
 					if sstChangedKeys && !extChangedKeys {
 						sstIter.SeekGE(MVCCKey{Key: extIter.UnsafeKey().Key})
 						sstOK, sstErr = sstIter.Valid()
+						if sstOK && extIter.UnsafeKey().Key.Compare(sstIter.UnsafeKey().Key) < 0 {
+							extIter.SeekGE(MVCCKey{Key: sstIter.UnsafeKey().Key})
+							extOK, extErr = extIter.Valid()
+						}
 					}
 					// Re-seek the ext iterator if the ext iterator changed keys and:
 					// 1) the SST iterator did not change keys, and we need to bring the ext

